### ISSUE TITLE
com.nextcloud.android.beta craches in IPV6PreferringDNS.lookup in constant restart loop (as long as wifi connected)

### ISSUE THREAD

**Issue by hlovdal** [nextcloud/android-library#594]
I have been running the beta android client the last couple of years basically with only one serious problem before, this it number two.

```log
03-12 16:56:41.174 16254 16375 E AndroidRuntime: FATAL EXCEPTION: Thread-5
03-12 16:56:41.174 16254 16375 E AndroidRuntime: Process: com.nextcloud.android.beta, PID: 16254
03-12 16:56:41.174 16254 16375 E AndroidRuntime: java.lang.UnsupportedOperationException
03-12 16:56:41.174 16254 16375 E AndroidRuntime:        at java.util.AbstractList.set(AbstractList.java:132)
03-12 16:56:41.174 16254 16375 E AndroidRuntime:        at java.util.AbstractList$ListItr.set(AbstractList.java:426)
03-12 16:56:41.174 16254 16375 E AndroidRuntime:        at java.util.Collections.sort(Collections.java:247)
03-12 16:56:41.174 16254 16375 E AndroidRuntime:        at com.nextcloud.common.IPV6PreferringDNS.lookup(IPV6PreferringDNS.java:30)
03-12 16:56:41.174 16254 16375 E AndroidRuntime:        at okhttp3.internal.connection.RouteSelector.resetNextInetSocketAddress(RouteSelector.kt:164)
03-12 16:56:41.174 16254 16375 E AndroidRuntime:        at okhttp3.internal.connection.RouteSelector.nextProxy(RouteSelector.kt:129)
03-12 16:56:41.174 16254 16375 E AndroidRuntime:        at okhttp3.internal.connection.RouteSelector.next(RouteSelector.kt:71)
03-12 16:56:41.174 16254 16375 E AndroidRuntime:        at okhttp3.internal.connection.ExchangeFinder.findConnection(ExchangeFinder.kt:205)
03-12 16:56:41.174 16254 16375 E AndroidRuntime:        at okhttp3.internal.connection.ExchangeFinder.findHealthyConnection(ExchangeFinder.kt:106)
03-12 16:56:41.174 16254 16375 E AndroidRuntime:        at okhttp3.internal.connection.ExchangeFinder.find(ExchangeFinder.kt:74)
03-12 16:56:41.174 16254 16375 E AndroidRuntime:        at okhttp3.internal.connection.RealCall.initExchange$okhttp(RealCall.kt:255)
03-12 16:56:41.174 16254 16375 E AndroidRuntime:        at okhttp3.internal.connection.ConnectInterceptor.intercept(ConnectInterceptor.kt:32)
03-12 16:56:41.174 16254 16375 E AndroidRuntime:        at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.kt:109)
03-12 16:56:41.174 16254 16375 E AndroidRuntime:        at okhttp3.internal.cache.CacheInterceptor.intercept(CacheInterceptor.kt:95)
03-12 16:56:41.174 16254 16375 E AndroidRuntime:        at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.kt:109)
03-12 16:56:41.174 16254 16375 E AndroidRuntime:        at okhttp3.internal.http.BridgeInterceptor.intercept(BridgeInterceptor.kt:83)
03-12 16:56:41.174 16254 16375 E AndroidRuntime:        at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.kt:109)
03-12 16:56:41.174 16254 16375 E AndroidRuntime:        at okhttp3.internal.http.RetryAndFollowUpInterceptor.intercept(RetryAndFollowUpInterceptor.kt:76)
03-12 16:56:41.174 16254 16375 E AndroidRuntime:        at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.kt:109)
03-12 16:56:41.174 16254 16375 E AndroidRuntime:        at okhttp3.internal.connection.RealCall.getResponseWithInterceptorChain$okhttp(RealCall.kt:201)
03-12 16:56:41.174 16254 16375 E AndroidRuntime:        at okhttp3.internal.connection.RealCall.execute(RealCall.kt:154)
03-12 16:56:41.174 16254 16375 E AndroidRuntime:        at com.nextcloud.common.OkHttpMethodBase.execute(OkHttpMethodBase.kt:171)
03-12 16:56:41.174 16254 16375 E AndroidRuntime:        at com.nextcloud.client.network.ConnectivityServiceImpl.isInternetWalled(ConnectivityServiceImpl.java:75)
03-12 16:56:41.174 16254 16375 E AndroidRuntime:        at com.owncloud.android.utils.FilesSyncHelper.lambda$restartJobsIfNeeded$0(FilesSyncHelper.java:207)
03-12 16:56:41.174 16254 16375 E AndroidRuntime:        at com.owncloud.android.utils.-$$Lambda$FilesSyncHelper$78U1WN_nzr5HiGIrRZMThSEamz8.run(lambda)
03-12 16:56:41.174 16254 16375 E AndroidRuntime:        at java.lang.Thread.run(Thread.java:761)
03-12 16:56:41.241   562 16383 I chatty  : uid=1000 system_server expire 1 line
03-12 16:56:41.268   562  6111 I ActivityManager: Process com.nextcloud.android.beta (pid 16254) has died
03-12 16:56:41.269   562  6111 D ActivityManager: cleanUpApplicationRecord -- 16254
03-12 16:56:41.269   562  6111 W ActivityManager: Scheduling restart of crashed service com.nextcloud.android.beta/androidx.work.impl.background.systemjob.SystemJobService in 1000ms
03-12 16:56:41.269   562  6111 W ActivityManager: Scheduling restart of crashed service com.nextcloud.android.beta/com.owncloud.android.services.AccountManagerService in 11000ms
03-12 16:56:41.269   562  6111 I ActivityManager: Killing 16260:com.nextcloud.android.beta:crash/u0a184 (adj 0): depends on provider com.nextcloud.android.beta/com.owncloud.android.providers.FileContentProvider in dying proc com.nextcloud.android.beta (adj 0)
03-12 16:56:41.315   562  2187 I WindowManager: WIN DEATH: Window{bf2c98c u0 com.nextcloud.android.beta/com.nextcloud.client.errorhandling.ShowErrorActivity}
03-12 16:56:41.325   562  8038 I ActivityManager: Start proc 16384:com.nextcloud.android.beta/u0a184 for service com.nextcloud.android.beta/com.owncloud.android.services.AccountManagerService
03-12 16:56:41.326   562  8294 D ActivityManager: cleanUpApplicationRecord -- 16260
03-12 16:56:41.327   562  8294 W ActivityManager: Force removing ActivityRecord{ef0a427 u0 com.nextcloud.android.beta/com.nextcloud.client.errorhandling.ShowErrorActivity t3751}: app died, no saved state
03-12 16:56:41.386   562   583 W WindowManager: Failed looking up window
```

Looking at the code, the function is quite simple

```java
        public List<InetAddress> lookup(String hostname) throws UnknownHostException {
            List<InetAddress> addresses = cache.get(hostname.toLowerCase(Locale.ROOT));

            if (addresses != null) {
                return addresses;
            }

            addresses = Dns.SYSTEM.lookup(hostname);
            Collections.sort(addresses, (address1, address2) -> {
                if(address1 instanceof Inet4Address) {
                    return 1;
                } else {
                    return -1;
                }
            });
            cache.put(hostname, addresses);

            return addresses;
        }
```

I am not sure what types `Dns.SYSTEM.lookup` returns, but looking into why `sort` can fail, it seems to be cause by missing `set` support on the elements.

https://stackoverflow.com/questions/53065894/collections-sort-throws-unsupportedoperationexception-while-sorting-a-list-after
https://stackoverflow.com/questions/21854353/why-does-collections-sort-throw-unsupported-operation-exception-while-sorting-by
https://stackoverflow.com/questions/55802573/abstractlist-unsupportedoperationexception-when-calling-sort-on-a-list

Probably you need to wrap the result in an ArrayList before attempting to sort?

**Comment by tobiasKaminsky** [comment]
We store addresses in a HashMap (L23 and L19), so this should be ok.

What device do you have?

**Comment by hlovdal** [comment]
It's a nexus 7, Android version 7.1.2 (14.1-20190208-NIGHTLY-flo).

**Comment by tobiasKaminsky** [comment]
I tried it with an official Android 7 emulator and did not have any problems…
Assuming from your version name, you are using LineageOS, maybe there are some missing parts…

**Comment by jbarnoin** [comment]
I just started getting the same issue, the callstack looks identical. I'm not sure what changed as the app worked recently on this device.
I'm not using LineageOS, this is a Samsung Galaxy Note 3, with a stock but admittedly old Android version (5.0). Still, the application was working fine and suddenly started getting into a crash loop on this callstack until I turn off all internet connectivity. This is extremely inconvenient as just having the app installed now makes this device unusable, with this crash report getting in the way of doing anything else and interrupting any other application every few seconds. I'm not sure why it suddenly started behaving this way today.

**Comment by tobiasKaminsky** [comment]
@jbarnoin have you updated the app?

**Comment by jbarnoin** [comment]
I'm quite sure the same version that used to work fine started behaving like this without an update in-between. I had been using it earlier in the day and it suddenly started happening, and I cannot determine why.
What I can say is that:
- The issue persisted after a reboot of the device
- I was enrolled in the Beta. I removed the beta and installed the regular version, and things worked fine. I then installed the beta again, and things seem fine again on the same device.

So... it's not clear how this is triggered, and I can't reproduce this any longer at the moment. It clearly doesn't impact everyone, however when it does happen it is pretty bad, I had to be fast to take back control of the device by managing to uninstall the app in-between crash report screens.
